### PR TITLE
[SPARK-10364][SQL] Support Parquet logical type TIMESTAMP_MILLIS

### PR DESCRIPTION
**Pull request by dilipbiswal**
## What changes were proposed in this pull request?

**Description** from JIRA

The TimestampType in Spark SQL is of microsecond precision. Ideally, we should convert Spark SQL timestamp values into Parquet TIMESTAMP_MICROS. But unfortunately parquet-mr hasn't supported it yet.
For the read path, we should be able to read TIMESTAMP_MILLIS Parquet values and pad a 0 microsecond part to read values.
For the write path, currently we are writing timestamps as INT96, similar to Impala and Hive. One alternative is that, we can have a separate SQL option to let users be able to write Spark SQL timestamp values as TIMESTAMP_MILLIS. Of course, in this way the microsecond part will be truncated.
## How was this patch tested?

Added new tests in ParquetQuerySuite and ParquetIOSuite

### CODE REVIEW

**Comment by davies** [comment]
Paruqet support both millisecond and microseconds:
```
TIMESTAMP_MILLIS

TIMESTAMP_MILLIS is used for a combined logical date and time type, with millisecond precision. It must annotate an int64 that stores the number of milliseconds from the Unix epoch, 00:00:00.000 on 1 January 1970, UTC.

TIMESTAMP_MICROS

TIMESTAMP_MICROS is used for a combined logical date and time type with microsecond precision. It must annotate an int64 that stores the number of microseconds from the Unix epoch, 00:00:00.000000 on 1 January 1970, UTC.
```

We should be able to read both of them, but write it as TIMESTAMP_MICROS by default?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -206,6 +206,22 @@ object DateTimeUtils {
     (day.toInt, micros * 1000L)
   }
 
+  /*
+   * Converts the timestamp to milliseconds since epoc. In spark timestamp values have microseconds
+   * precision, so this conversion is lossy.
+   */
+  def toMillis(us: SQLTimestamp): Long = {
+    val julian_us = us + JULIAN_DAY_OF_EPOCH * MICROS_PER_DAY
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Where does this come from? It's different than that in Parquet doc (since epoch)

Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @davies Thank you for your  comments Davies. Per Lian's comment in the jira,
https://issues.apache.org/jira/browse/SPARK-8824
parquet 1.8 which is what we are using currently does not have support for TIMESTAMP_MICROS yet. He suggested we implement TIMESTAMP_MILLIS for now.

Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @davies On your second comment , let me please get back to you.

Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @davies Many thanks !! You are right. We need to write it as milliseconds since epoc.  I will send a update .. Thanks again !!

**Comment by dilipbiswal** [comment]
retest this please

**Comment by liancheng** [comment]
@davies Unfortunately parquet-mr 1.8.1, which is used by the current master, hadn't included `TIMESTAMP_MICROS` yet. To be more specific, `OriginalType` in parquet-mr 1.8.1 doesn't include `TIMESTAMP_MICROS` as a member. So I think only support `TIMESTAMP_MILLIS` is reasonable here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala [hunk 2]
```diff
@@ -312,6 +316,9 @@ class ParquetFileFormat
     hadoopConf.setBoolean(
       SQLConf.PARQUET_INT96_AS_TIMESTAMP.key,
       sparkSession.sessionState.conf.isParquetINT96AsTimestamp)
+    hadoopConf.setBoolean(
+      SQLConf.PARQUET_INT64_AS_TIMESTAMP_MILLIS.key,
+      sparkSession.sessionState.conf.isParquetINT96AsTimestamp)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: `isParquetINT64AsTimestampMillis` here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 3]
```diff
@@ -206,6 +207,30 @@ object DateTimeUtils {
...
+    if (us < 0 && (us % MILLIS_PER_SECOND < 0)) {
+      millis = millis - 1
+    }
+    millis
+  }
+
+  /*
+   * Converts millseconds since epoc to SQLTimestamp.
+   */
+  def fromMillis(millis: SQLTimestamp): Long = {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Return type should be `SQLTimestamp`? Input type should be `Long`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala (hunk 2 above)
Comment (dilipbiswal) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFileFormat.scala]: @viirya Thank you. will change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 3 above)
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @viirya Will make the change. Thanks !!

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 4]
```diff
@@ -362,7 +363,15 @@ private void readLongBatch(int rowId, int num, ColumnVector column) throws IOExc
     if (column.dataType() == DataTypes.LongType ||
         DecimalType.is64BitDecimalType(column.dataType())) {
       defColumn.readLongs(
-          num, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
+              num, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: For vectorized reader, I think we should also add `TimestampType` support for `INT64` in `decodeDictionaryIds`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala [hunk 5]
```diff
@@ -356,6 +363,9 @@ private[parquet] class ParquetSchemaConverter(
       case DateType =>
         Types.primitive(INT32, repetition).as(DATE).named(field.name)
 
+      case TimestampType if writeTimestampInMillis =>
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaConverter.scala]: We have a block of comments and codes for `TimestampType` below. Can we move this branch in the block? And we should add few comments about this change.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala [hunk 6]
```diff
@@ -164,6 +165,63 @@ class ParquetQuerySuite extends QueryTest with ParquetTest with SharedSQLContext
...
+        sql("select * from ts"),
+        Seq(
+          Row(1, Timestamp.valueOf("2016-01-01 10:11:12.123456")),
+          Row(2, null),
+          Row(3, Timestamp.valueOf("1965-01-01 10:11:12.123456"))))
+    }
+
+    // When written as TIMESTAMP_MILLIS, the microseconds are truncated.
+    withTable("ts") {
+      withSQLConf(SQLConf.PARQUET_INT64_AS_TIMESTAMP_MILLIS.key -> "true") {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala]: Can you add a test that can we read a timestamp field written with `SQLConf.PARQUET_INT64_AS_TIMESTAMP_MILLIS` as `true`, but reading it with `SQLConf.PARQUET_INT64_AS_TIMESTAMP_MILLIS` as `false`?

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java (hunk 4 above)
Comment (dilipbiswal) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: @viirya Thanks Simon. I have made the changes.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala (hunk 6 above)
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetQuerySuite.scala]: @viirya Done

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 7]
```diff
@@ -157,7 +158,8 @@ void readBatch(int total, ColumnVector column) throws IOException {
             num, dictionaryIds, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
         if (column.hasDictionary() || (rowId == 0 &&
             (descriptor.getType() == PrimitiveType.PrimitiveTypeName.INT32 ||
-            descriptor.getType() == PrimitiveType.PrimitiveTypeName.INT64 ||
+            (descriptor.getType() == PrimitiveType.PrimitiveTypeName.INT64  &&
+               column.dataType() != DataTypes.TimestampType) ||
```
Comment (viirya) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: Why this check?

Comment (dilipbiswal) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: We need to convert the time back to micros and in case of lazy decoding, we don't get that chance ?

**Comment by viirya** [comment]
LGTM. see if @davies @liancheng have other comments about this.

**Comment by saulshanabrook** [comment]
Would it be helpful to submit a new PR with the conflicts resolved? If not, what are the next steps for this issue?

**Comment by dilipbiswal** [comment]
@saulshanabrook Hello, thanks for your comment. Currently, i am waiting for feedback from @liancheng and @davies. Perhaps this is not a priority now. I will try to resolve the conflicts and push in any case.

**Comment by gatorsmile** [comment]
Also cc @ueshin and @squito

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 8]
```diff
@@ -237,6 +238,30 @@ object DateTimeUtils {
...
+  def toMillis(us: SQLTimestamp): Long = {
+    var millis = us / 1000L
+
+    // When the timestamp is negative i.e before 1970, we need to adjust the millseconds portion.
+    // Example - 1965-01-01 10:11:12.123456 is represented as (-157700927876544) in micro precision.
+    // In millis precision the above needs to be represented as (-157700927877)
+
+    if (us < 0 && (us % MILLIS_PER_SECOND < 0)) {
+      millis = millis - 1
+    }
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Can't we use `Math.floor()` here as the same as `millisToDays`?

**Review by ueshin — COMMENTED** [review]
LGTM, except for minor comments.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala [hunk 9]
```diff
@@ -965,6 +972,18 @@ class ParquetSchemaSuite extends ParquetSchemaTest {
     int96AsTimestamp = true,
     writeLegacyParquetFormat = true)
 
+  testSchema(
+    "Timestmp written and read as INT64 with TIMESTAMP_MILLIS",
```
Comment (dilipbiswal) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetSchemaSuite.scala]: @ueshin Thanks. Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 8 above)
Comment (dilipbiswal) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @ueshin Thanks a lot. I have made the change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 10]
```diff
@@ -237,6 +238,24 @@ object DateTimeUtils {
     (day.toInt, micros * 1000L)
   }
 
+  /*
+   * Converts the timestamp to milliseconds since epoc. In spark timestamp values have microseconds
```
Comment (squito) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: typo: epoch

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java [hunk 11]
```diff
@@ -362,7 +375,15 @@ private void readLongBatch(int rowId, int num, ColumnVector column) throws IOExc
     if (column.dataType() == DataTypes.LongType ||
         DecimalType.is64BitDecimalType(column.dataType())) {
       defColumn.readLongs(
-          num, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
+              num, column, rowId, maxDefLevel, (VectorizedValuesReader) dataColumn);
```
Comment (squito) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: why this change in indentation?  if anything, looks like it should be indented *less* than the original.

**Review by squito — COMMENTED** [review]
I just found a couple of tiny nits.  I will defer to others on this change.

I guess we'll want to add something similar to SPARK-12297 for int64 as well eventually, but I dont' think they need to go in together, especially as both are in-flight right now.

File: sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java (hunk 11 above)
Comment (dilipbiswal) [inline sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedColumnReader.java]: @squito Thank you for reviewing. I have fixed the indentation.

**Comment by ueshin** [comment]
Thanks! Merging to master.

**Comment by dilipbiswal** [comment]
Thanks a lot @ueshin @viirya @gatorsmile

**Comment by cloud-fan** [comment]
Hi @dilipbiswal , do you mind to share how you generate the testing parquet file? thanks!

**Comment by dilipbiswal** [comment]
@cloud-fan Hi Wenchen, its been a while .. i am trying my best to recollect. I think once i had the write code implemented in spark, i used it to produce files. Depending on the data, parquet uses different encoding (plain or dictionary). I examined the encodings using parquet tools. I produced two files and used the merge option in the tools to merge them in one file. This is to the best of my recollection :-)

**Comment by dilipbiswal** [comment]
@cloud-fan fyi - https://github.com/apache/parquet-mr/tree/master/parquet-tools

**Comment by cloud-fan** [comment]
great, thanks!
